### ISSUE TITLE
Liquigraph and neo4j upgrade: how to handle existing migrations

### ISSUE THREAD

**Issue by gonzalad** [liquibase/liquigraph#382]
Hello,

We're going to upgrade from neo4j 3.x to neo4j 4.x.

Some of our liquigraph migrations use syntax that is no longer supported in Neo4j 4.x (see https://neo4j.com/docs/cypher-manual/current/deprecations-additions-removals-compatibility/).

For instance, I'll need to rewrite:

```
MATCH (n)-[:A|:B|:C {foo: 'bar'}]-() RETURN n
```

to

```
MATCH (n)-[:A|B|C {foo: 'bar'}]-() RETURN n
```

But rewritting migrations means that the MD5 checksum will change and liquigraph will throw an error next time itis executed.

What is the best way to handle this kind of situation ?

Looking at a possible solution (don't know it it's the right way), I've seen that liquibase supports a clearCheckSums command that could do the trick (https://www.liquibase.org/documentation/command_line.html).

Do we have support for that in liquigraph ?

Otherwise, is it something interesting enough for a contribution ?

Thanks,
Adrian

**Comment by fbiville** [comment]
Hi @gonzalad, thanks for the report, that's a great question!
There is no official support for this situation and there should be!

In the meantime, I believe you can follow these steps:

1. effectively disable the faulty migration with a precondition (something like `RETURN false AS result`). That way, you can keep it around, comment why it's now disabled instead of removing it. Due to a bug that yet needs fixing, adding a precondition to a changeset won't alter its checksum (so you can exploit this at your advantage for now! - see #363, #364 and #365)
2. now that the faulty query is disabled, you can add the replacement. However, this new changeset should be executed if and only if the previous one has never been executed. A precondition is needed here again. While it's generally not a good idea (because this relies on Liquigraph internal structures), here you can check whether the Liquigraph history graph contains the faulty changeset has not been executed, something like:
```
MATCH (changelog:__LiquigraphChangelog) 
OPTIONAL MATCH (changelog)<-[:EXECUTED_WITHIN_CHANGELOG]-(changeset:__LiquigraphChangeset {id: __CHANGESET__ID__, author: __CHANGESET__AUTHOR__})
RETURN COALESCE(ID(changeset), "not-in-db") = "not-in-db" AS result
```

This will work for the time being.
Unfortunately, this may break in the future as this relies on internal structure that may evolve but I think that's the best you can do until there is official support for that kind of situation.

**Comment by gonzalad** [comment]
Thanks very much for your help @fbiville !

I'll try your suggestion for one or 2 queries (don't know how much I will have ATM).

At the same time, I'll try to hack some support for a clearCheckSums and make a PR.
Not sure that's the way you'll want to handle this kind of case in the future of liquigraph, but if it is and if the PR looks good enough, then everyone wins :)

Thanks,
Adrian

**Comment by fbiville** [comment]
Awesome!

The change may be quite big and spread over different modules. May I suggest you focus the PR on the core module first?

You'd need to introduce a new method on `org.liquigraph.core.api.Liquigraph` (e.g. `clearChecksums`) and you can probably write the Cypher logic in a new class in the package `org.liquigraph.core.io` (make sure you rely on `GraphJdbcConnector` to connect to the database, as it already implements a lock mechanism to prevent concurrent execution of Liquigraph).

Then, you'll need to update `org.liquigraph.core.io.ChangelogGraphReader#read` to make sure to compute any missing checksum.

Happy to help, feel free to ping me if you have more questions! (there is a channel called `#using-liquigraph-` in neo4j-users.slack.com)

**Comment by gonzalad** [comment]
Hi @fbiville, access to neo4j-users.slack.com seems to be restricted for individuals having an email ending with @neotechnology.com, @graphaware.com, @structr.com ou @neo4j.com.

So I won't be able to join you on slack.

In the meantime, I've just created the PR https://github.com/liquigraph/liquigraph/pull/385.

Very rough, I coded it before your last reply in this thread.

The quality is quite poor (sorry for that), I didn't want to make premature and unecessary refactorings and just kept it bare bones atm.

If you have any kind of feedback on the PR, I'll gladly make the changes.

Thanks,
Adrian
